lib/licence_location_identifier.rb

```diff
@@ -5,27 +5,57 @@ class LicenceLocationIdentifier < LocationIdentifier
   # to find the most appropriate council from the "providing tier"
   # if not, then return the closest authority available.
   def self.find_slug(areas, artefact)
-    councils = areas.map do |area|
-      [area["type"], area["codes"] && area["codes"]["govuk_slug"]]
-    end
+    new(areas, artefact).find_slug
+  end
 
-    if artefact and providing_tier = artefact["details"].try(:[], "licence").try(:[], "local_service").try(:[], "providing_tier")
-      councils_by_tier = Hash[councils.collect do |k, v|
-        [self.identify_tier(k), v]
-      end]
+  attr_reader :areas, :artefact
 
-      appropriate_authority = providing_tier.map do |tier|
-        councils_by_tier[tier]
-      end.compact.first
+  def initialize(areas, artefact)
+    @areas = areas
+    @artefact = artefact
+  end
 
-      return appropriate_authority unless appropriate_authority.nil?
-    end
+  def find_slug
+    matching_authority_by_tier_slug || matching_authority_by_area_type_slug
+  end
 
-    authorities = Hash[councils]
-    self.authority_types.each do |type|
-      return authorities[type] unless authorities[type].nil?
+private
+
+  def matching_authority_by_tier_slug
+    matching_authority_by_tier.try(:[], "codes").try(:[], "govuk_slug")
+  end
+
+  def matching_authority_by_tier
+    return nil unless service_providing_tiers
+
+    service_providing_tiers.each do |tier|
+      areas.each do |area|
+        return area if tier == LocationIdentifier.identify_tier(area["type"])
+      end
     end
+    nil
+  end
+
+  def service_providing_tiers
+    return unless artefact
 
+    artefact["details"].try(:[], "licence").try(:[], "local_service").try(:[], "providing_tier")
+  end
+
+  def matching_authority_by_area_type_slug
+    matching_authority_by_area_type.try(:[], "codes").try(:[], "govuk_slug")
+  end
+
+  def matching_authority_by_area_type
+    LocationIdentifier.authority_types.each do |type|
```

Comment:
The inner `each` could probably be written with `detect`.

```ruby
LocationIdentifier.authority_types.each do |type|
  found = areas.detect { |area| area["type"] == type && area_has_slug?(area) }
  return found if found
end
```

But I'm not hugely convinced it's much clearer.  I reckon these nested `each` loops are probably ok.